proposals/0030-core-image-story.md

```diff
@@ -0,0 +1,402 @@
+
+# Background & Motivation
+
+The currently available `@astrojs/image` is great, however, users ultimately find it very confusing for multiple reasons:
+
+- Error messages are confusing and unhelpful
+- Types historically haven't been super great
+- Missing documentation
+- Due to not being a core integration, it requires manual setup
+- Partly due to the previous points, but also for other reasons, it wasn't always clear for users how the integration behaved.
+- Hard to use in Markdown / MDX.
+
+In this RFC, we'd like to outline a plan / API for a core story for images. The main motivation is to create a story that feels fitting to the Astro core, in that it's easy to understand, intuitive to use, and extendable while also being kind and respectful of the user.
+
+# Goals
+
+- Make an image component that is easy to use and intuitive to understand.
+- Make it easy to use optimized images in Markdown, MDX and future formats.
+- Good, core-worthy, DX with awesome error messages, good types, good documentation
+- No more integration to install!
+
+# Non-Goals of this RFC
+
+- Advanced usage in Markdown (ability to set width, height, quality etc)
+- Using optimized images inside client-side framework components
+- Automatic generation of `srcset` in the integrated services
+- Automatic `loading="eager"` for above the fold images
+- Placeholders generation
+- Background generation
+- Picture component
+- Optimizing & resizing remote images
+- `.svg` support
+
+To be clear, we hope to tackle many of those points in the future, in separate, more precise RFCs. Images are hard.
+
+# Detailed Design
+
+## `astro:assets`
+
+A new virtual module will be exported from a Vite plugin (similar to `astro:content`) exposing the different tools the user can access.
+
+- `Image` (see [Image Component](#image-component-1))
+- `getImage` (see [JavaScript API](#javascript-api))
+- `getImageService` (see [Image Services](#image-services))
+
+We choose `astro:assets` over `astro:image` on purpose, as to make it intuitive that more things might get exposed from there over time.
+
+## `src/assets`
+
+A new reserved folders for assets. Assets do not have to live in this folder, however putting them in this folder will unlock several benefits. Notably, we consider this folder to be a safe place for git-based CMSes to put any uploaded assets in.
+
+Through an included alias `~/assets` pointing to `src/assets`, it'll be easy for users to refer to any uploaded assets there.
+
+### Content Collections integration
+
+It is fairly common for one of the property of a Markdown piece of content to need to be a reference to an asset (think, cover image for an article, picture for an author etc).
+
+In tandem with the `src/assets` folder, we'd like to introduce a way for users to specify that a specific property needs to refer to a valid asset from the `src/assets` folder:
+
+```ts
+import { asset, defineCollection, z } from "astro:content";
```

Comment:
(bikeshed comment, not blocking my approval on this RFC)

Any thoughts on matching `asset` naming? My brain says that `import {asset}` is too generic, but I see that you're trying to match the format of `z.asset()` which makes sense. `import {assetSchema}` is less generic, but now feels so different from `z.asset()`.

What about `import {a} from 'astro:content';`? Then this snippet would look like:

```ts
import { defineCollection, a, z } from "astro:content";
const blogCollection = defineCollection({
  schema: z.object({
    title: z.string(),
    image: a.asset({ width: 1120, height: 1120 }),
  }),
});
```

Or, another idea that's probably less controversial:


```ts
import { defineCollection, schemas, z } from "astro:content";
const blogCollection = defineCollection({
  schema: z.object({
    title: z.string(),
    image: schemas.asset({ width: 1120, height: 1120 }),
  }),
});
```

IDK, maybe too early to be thinking about this since it's the first custom schema we're shipping. Like I said, no need to block on this, and also happy to revisit when we add our second or third custom schema.

Reply:
I can't say I have any particular thoughts on this. I like the `schemas.asset` one though, but maybe `schema` is a bit too foreign for users not used to Zod?